spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,120 @@
+Instead of client libraries invoking `pkl server`, `pkl eval` invocations, when configured with an external reader, will run the reader executable as a subprocess.
+External readers can provide any number of resource and/or module schemes and will configured as an executable and list of arguments via:
+
+* CLI flag
+* PklProject property
+* Java API
+* Message passing API
+
+
+[source,pkl]
+----
+/// May be specified as an URI to a readable resource, eg. `file:`, `https:`, or `package:`
+/// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+executable: String
+
+/// Command line arguments that will be passed to the reader process
+arguments: Listing<String>
+----
+
+=== Example
+
+Consider this module:
+
+[source,pkl]
+----
+username = "john_appleseed"
+
+email = read("ldap://ds.example.com:389/dc=example,dc=com?mail?sub?(uid=\(username))").text
+----
+
+Pkl doesn't implement the `ldap:` resource URI scheme natively, but an external reader can provide it.
+Assuming a hypothetical `pkl-ldap` binary implementing the external reader protocol and the `ldap:` scheme is in the `$PATH`, this module can be evaluated as:
+
+[source,text]
+----
+$ pkl eval <module> --external-reader=pkl-ldap --allowed-resources prop:,ldap:
+username = "john_appleseed"
+email = "appleseed@example.com"
+----
+
+For security reasons, schemes provided by external readers are not allowed by default and clients must explicitly allow the desired scheme(s).
+In this example, the external reader may provide both `ldap:` and `ldaps:` schemes.
+If a codebase or user wishes to enforce that only secure LDAP connections are made, the `ldaps:` scheme may be allowed while `ldap:` is not.
+
+== Detailed design
+
+To avoid terminology confusion with the existing language binding message passing model, the `pkl` process will remain known as the "server" and the reader process will be known as the "client", despite all requests in this proposal being initiated by the `pkl` process.
+
+* Message Passing API
+** Add DiscoverReadersRequest
+** Add DiscoverReadersResponse
+** CreateEvaluatorRequest
+*** `externalReaders`
+** Pull into a new pkl-msgapi(?) package shared by pkl-code and pkl-server
+*** Message*.kt files at least
+
+* Java API
+** New `ExternalModuleKeyFactory` type
+*** Entrypoints (CLI, Server) construct based on `externalReaders`
+*** On (initialization? explicit startup?), launch subprocess
+*** send each DiscoverReadersRequest, await DiscoverReadersResponse (timeout? 1s?)
+
+* stdlib
+** `pkl:EvaluatorSettings` (used in `pkl:Project`)
+*** `externalReaders: Listing<ExternalReader>`
+
+* CLI
+** `--external-reader`
+*** use `clikt.multiple`
+*** may be passed as a space-separated string where the first element becomes `executable` and any remainder becomes `arguments`
+*** should this be shlex'd? or just split on spaces?
+
+* Language Bindings
+** Add `ExternalReaderRuntime`
```

Comment:
What is `ExternalReaderRuntime`?

Reply:
Open to suggestions on the name, but this would be the functionality in the language binding libraries that runs the subprocess-side message passing actor, eg. in pkl-go (API obviously not final)
```go
type ExternalReaderRuntime interface {
  Run()
  Close()
}

func NewExternalReaderRuntime(resourceReaders []ResourceReader, moduleReaders []ModuleReader) ExternalReaderRuntime {
  // ...
}
```

Reply:
Ah, I see.

It seems a bit strange to expect clients themselves to spawn a sub-process? If we support this API in server mode, I'd expect pkl to spawn these external readers, rather than the clients, because:

* Pkl has all the information it needs already (what executable to spawn, what arguments to pass)
* Implementations can diverge between language bindings

Reply:
I think I might not have been clear? `ExternalReaderRuntime` would be provided by language binding libraries as a means to easily implement the "client" side of external reader binaries. It would be invoked by pkl itself and not the server client.

Here's what a mostly complete entrypoint to a simple `pkl-ldap` implementation might look like in Go:
```go
func main() {
  ldapReader := ldapReader{scheme: "ldap"} // implementation elided, not important to the point
  ldapsReader := ldapReader{scheme: "ldaps"}
  runtime := pkl.NewExternalReaderRuntime([]ResourceReader{ldapReader, ldapsReader}, nil)
  runtime.Run()
}
```

Building this into the existing language binding libraries is a convenience because the message-passing infrastructure already exists there. I can't think of a situation where a pkl server client and an external reader client would exist in the same application, as it would make more sense to use a client reader in the pkl server client directly.

Reply:
Oooh I see--this is code for the _reader_ itself, rather than from an application calling into Pkl.